src/lsst/cmservice/common/panda.py

```diff
@@ -0,0 +1,145 @@
+    determined from the panda url and the oidc VO.
+    """
+
+    # If no panda configuration exists in the application, return early
+    if config.panda.url is None:
+        return None
+
+    # If a token has been added to the configuration object, use it instead of
+    # loading one from disk
+    try:
+        if config.panda.refresh_token is None:
+            token_data = json.loads((Path(config.panda.config_root) / ".token").read_text())
+            config.panda.id_token = token_data["id_token"]
+            config.panda.refresh_token = token_data["refresh_token"]
+    except (FileNotFoundError, json.JSONDecodeError):
+        logger.exception()
+        return None
+
+    now_utc = datetime.datetime.now(datetime.UTC)
+
+    # Determine whether the token should be renewed
+    # The token expiry time is part of the encoded token
+    try:
+        if config.panda.token_expiry is None:
+            decoded_token = decode_id_token(config.panda.id_token)
+            config.panda.token_expiry = float(decoded_token["exp"])  # type: ignore
+        if TYPE_CHECKING:
+            # the validation machinery of the pyantic field handles conversion
+            # from float to datetime.
+            assert isinstance(config.panda.token_expiry, datetime.datetime)
+    except Exception:
+        # NOTE: this should generally be an AttributeError but the 3rdparty
+        #        function may change its operation.
+        # If current id_token is None or otherwise not decodable, we will get a
+        # new one from the refresh operation
+        logger.exception()
+        config.panda.token_expiry = now_utc
+
+    if (config.panda.token_expiry - now_utc) < datetime.timedelta(config.panda.renew_after):
+        if config.panda.auth_config_url is None:
+            logger.error("There is no PanDA auth config url known to the service, cannot refresh token.")
+            logger.warning("The current PanDA id token may be invalid.")
+            return config.panda.id_token
+
+        try:
+            # TODO it is probably safe to cache these response tokens
+            with http_client() as session:
+                auth_config_response = session.get(config.panda.auth_config_url)
+                auth_config_response.raise_for_status()
+                panda_auth_config = auth_config_response.json()
+
+                token_response = session.get(panda_auth_config["oidc_config_url"])
+                token_response.raise_for_status()
+                token_endpoint = token_response.json()["token_endpoint"]
+
+                data = dict(
+                    client_id=panda_auth_config["client_id"],
+                    client_secret=panda_auth_config["client_secret"],
+                    grant_type="refresh_token",
+                    refresh_token=config.panda.refresh_token,
+                )
+
+            _ = refresh_panda_token(token_endpoint, data)
+        except httpx.HTTPStatusError:
```

Comment:
I could imagine wanting to catch json parsing errors or missing keys in the Panda responses somewhere too, maybe here?